### ISSUE TITLE
Clean up metadata info for timeout requests in NetworkClient

### ISSUE THREAD

**Issue by nsivabalan** [linkedin/ambry#325]
We have two time out for a request. One at the networkClient layer where in we timeout incase we couldnt get a connection. Another one is at the OperationManager layer where in we timeout if we dont get a response after sometime(in other words, we haven't got the response in time even though the request was sent out). In later cases, the operationManager don't inform the NetworkClient regarding the timed outones. NetworkClient might be holding some unnecessary data for those requests even though its a no op. So, creating this ticket to do those cleanups.

**Comment by pnarayanan** [comment]
Not sure what this is about. `NetworkClient` does not hold on to any information regarding a request forever. Requests are either handed over to the `selector` (if there is a connection available) or timed out (if they remain in the queue for too long). When they time out, a `ResponseInfo` is constructed and returned to the caller.

I don't see an issue here.

**Comment by pnarayanan** [comment]
@nsivabalan could you either close this or provide a concrete reason to keep this?

**Comment by pnarayanan** [comment]
@nsivabalan ping.

**Comment by nsivabalan** [comment]
I haven't went into all the details. But to my knowledge, connectionIdToRequestInFlight will contain in flight requests. Lets say the Manager times-out for this operation(which has sent out this request) and closes the operation, how will the NetworkClient know that this(connectionIdToRequestInFlight) needs to be cleaned up.

**Comment by pnarayanan** [comment]
See my earlier comment.

This is not how the NetworkClient works. It takes in requests. Once it get them it does not allow for "canceling" requests. As for the callers, a request given to the NetworkClient is as good as sent.

Similarly, if the selector is provided a NetworkSend, the selector does not care if the caller later changes its mind. Anything handed over to the selector is as good as sent out over the network.

Basically, once you hand over a request or response to the Network layer it is as good as sent out over the network. You can't change your mind. This is true in general for any system where the network layer is cleanly separated from operation layers, not only for Ambry.

**Comment by vgkholla** [comment]
If I understand this issue correctly, the problem as stated might be a non-problem. The `NonBlockingRouter` (which is what I assume you mean by "managers") should not have to tell the `NetworkClient` anything.

Also, I'm not sure why the question of "cleaning" up arises at all considering that for the `NetworkClient`, the request is actually *still* in flight. The actual problem here, I think, is to decide whether we want a request timeout at the `NetworkClient`. For now, I don't see the utility of doing so because the connection is useless anyway - we can't send any more requests using the connection unless we receive a response for the previous one.  

To find out if this is a problem, we can track the size of the `connectionIdToRequestInFlight` map using a `Gauge` and also record how long each request spends in the map ("age"). In addition to removals in the happy path, we remove everything from the map even on `close()` and record the age of each entry. The 99th percentile will then tell us if something has been stuck in the map forever.  

What do you guys think?

**Comment by pnarayanan** [comment]
The `NetworkClient` either times out trying to checkout a connection to send the request, or simply hands over the request to the `Selector`. In addition, today it has to make sure that the connection does not get reused until a response is received for that request. We could add metrics to see how long a connection remains "in-use", sure. Siva probably has that already in #331.

The `NetworkClient` is just a liaison between the router and the selector. The `Selector` is the one that should ensure, if at all, that the request gets sent out or times out, as it is the one that should "disconnect" a connection.

**Comment by vgkholla** [comment]
Agreed.

**Comment by nsivabalan** [comment]
yep, I agree as well.

On Wed, May 25, 2016 at 2:59 PM, Gopalakrishna Holla <
notifications@github.com> wrote:

> Agreed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/linkedin/ambry/issues/325#issuecomment-221720877>
>



-- 
-Sivabalan
Data Infrastructure Engineer

**Comment by vgkholla** [comment]
Will close this - as discussed any requests that need to be timed out should be done at the `Selector` if we decide to go that route.

A helpful [link] (http://stackoverflow.com/a/17341089/5501399) on SO.

One of the comments on the answer
> Sure but you can't do that directly in case 1. You would have to keep track of the last read time for each channel and manipulate the select timeout so that the least recently read channel's timeout will expire if nothing happens, check all channels for timeouts, etc.
